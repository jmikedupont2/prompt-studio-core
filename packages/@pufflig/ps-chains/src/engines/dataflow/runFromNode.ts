import { ParamValue } from "@pufflig/ps-types";
import { Chain, NodeState, RunOptions } from "../../types";
import { resolveVariables } from "./utils/resolveVariables";
import { applyDefaultInputs, getEdgeMap, mapOutputToInput } from "./utils/utils";

/**
 * This function updates the input state of a node and executes the node on the new input.
 * Read all target nodes and update their input state.
 * Repeat until there are no more target nodes.
 *
 * @param nodeId id of the node to run
 * @param input input to update the node inputs with, empty object if existing inputs should be used
 * @param chain current definition and state of the chain
 * @param runOptions functions to call when a node state is updated or an error occurs
 */
export async function runFromNode(
  nodeId: string,
  input: Record<string, ParamValue>,
  chain: Chain,
  runOptions?: RunOptions
) {
  // avoid hanging on loops by keeping track of the nodes that have been visited
  // current implementation allows a node to be visited only as many times as it has inputs
  const visitedNodes: Record<string, number> = {};
  let chainState: Record<string, NodeState> = chain.state;

  async function _runFromNode(
    nodeId: string,
    input: Record<string, ParamValue>,
    chain: Chain,
    runOptions?: RunOptions,
    isChild = false
  ) {
    const nodeConfig = chain.definition.nodes[nodeId];
    const nodeDefinition = chain.nodeTypes[nodeConfig?.type];
    if (!nodeConfig || !nodeDefinition) {
      throw new Error(`Definition for node ${nodeId} not found`);
    }

    // parse the input
    const nodeState = chainState[nodeId];
    const prevInput = nodeState?.input;
    const parsedInput = await nodeDefinition.parseInput(input, prevInput);

    // a node can only be visited as many times as it has connected inputs
    const inputEdges = Object.values(chain.definition.edges).filter((e) => e.target === nodeId);
    const numInputs = inputEdges.length;

    if (isChild && (visitedNodes[nodeId] || 0) > numInputs) {
      return;
    }

    visitedNodes[nodeId] = visitedNodes[nodeId] ? visitedNodes[nodeId] + 1 : 1;

    const newInput = { ...applyDefaultInputs(prevInput, nodeDefinition), ...parsedInput };
    const newState: NodeState = { ...nodeState, input: newInput };

    let newChainState = { ...chainState, [nodeId]: newState };
    chainState = newChainState;

    runOptions?.onNodeInputUpdate?.(nodeId, newState);

    // do not run nodes in the chain that have autorun=false
    // except for the root node
    if (isChild) {
      if (nodeConfig.autorun === false) {
        return;
      }
    }

    // get all target nodes of the node we updated the input of
    // avoid running the node if it has no target nodes
    const targetNodes = Object.values(chain.definition.edges).filter((e) => e.source === nodeId);
    if (targetNodes.length === 0) {
      return;
    }

    let resolvedInput = newInput;
    try {
      resolvedInput = await resolveVariables(newInput, runOptions?.resolveReferences || (async (i) => i));
    } catch (error) {
      runOptions?.onNodeRunError?.(nodeId, error as Error);
      return;
    }

    // run the node we updated the input of
    let res;
    try {
      // resolve references in the input
      res = await nodeDefinition.execute(resolvedInput, prevInput);
      runOptions?.onNodeRunComplete?.(nodeId, res);
      // break the chain if the node returns null
      if (res === null) return;
    } catch (error) {
      // if running the node fails, stop the chain
      runOptions?.onNodeRunError?.(nodeId, error as Error);
      return;
    }

    const editedKeys = Object.keys(res);

    // ignore target nodes that are not connected to the values that have been edited
    const targetNodeIds = targetNodes.filter((e) => editedKeys.includes(e.sourceHandle)).map((e) => e.target);

    // update the state of all target nodes
    for (const targetNodeId of targetNodeIds) {
      // edges between this node and the target node
      const edges = Object.values(chain.definition.edges).filter(
        (e) => e.source === nodeId && e.target === targetNodeId
      );
      const edgeMap = getEdgeMap(edges);
      const mappedInput = mapOutputToInput(res, edgeMap);
      await _runFromNode(targetNodeId, mappedInput, chain, runOptions, true);
    }
  }

  await _runFromNode(nodeId, input, chain, runOptions);

  return chainState;
}
