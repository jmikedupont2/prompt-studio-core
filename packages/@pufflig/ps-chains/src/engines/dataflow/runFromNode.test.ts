import {
  autorunExample,
  autorunRootExample,
  missingStates,
  multiInput,
  simpleChain,
  simpleLoop,
  singleNodeChain,
} from "../../mocks/chains";
import { runFromNode } from "./runFromNode";

test("set input for single node in the editor", async () => {
  const onNodeInputUpdate = jest.fn();
  const res = await runFromNode("1", { template: "Hello World" }, singleNodeChain, {
    onNodeInputUpdate,
    onNodeRunError: () => {},
  });
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(1);
  expect(res).toMatchSnapshot();
});

test("set input for several nodes in the editor", async () => {
  const onNodeInputUpdate = jest.fn();
  const res = await runFromNode("n1", { template: "Hello {{World}}" }, simpleChain, {
    onNodeInputUpdate,
    onNodeRunError: () => {},
  });
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(3);
  expect(res).toMatchSnapshot();
});

test("ignore autorun=false flag on root node", async () => {
  const onNodeInputUpdate = jest.fn();
  const res = await runFromNode("n1", { template: "Hello {{World}}" }, autorunRootExample, {
    onNodeInputUpdate,
    onNodeRunError: () => {},
  });
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(3);
  expect(res).toMatchSnapshot();
});

test("nodes with autorun=false down the chain do not update the inputs of their targets", async () => {
  const onNodeInputUpdate = jest.fn();
  const res = await runFromNode("n1", { template: "Hello {{World}}" }, autorunExample, {
    onNodeInputUpdate,
    onNodeRunError: () => {},
  });
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(2);
  expect(res).toMatchSnapshot();
});

test("avoid hanging on loops", async () => {
  const onNodeInputUpdate = jest.fn();
  const res = await runFromNode("n1", { template: "Hello {{World}}" }, simpleLoop, {
    onNodeInputUpdate,
    onNodeRunError: () => {},
  });
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(2);
  expect(res).toMatchSnapshot();
});

test("can handle missing states", async () => {
  const onNodeInputUpdate = jest.fn();
  const res = await runFromNode("n2", { template: "Hello {{World}}" }, missingStates, {
    onNodeInputUpdate,
    onNodeRunError: () => {},
  });
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(4);
  expect(res).toMatchSnapshot();
});

test("can handle input from multiple origins", async () => {
  const onNodeInputUpdate = jest.fn();
  const res = await runFromNode("n1", { template: "Hello {{World}}" }, multiInput, {
    onNodeInputUpdate,
    onNodeRunError: () => {},
  });
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(4);
  expect(res).toMatchSnapshot();
});

test("update array variables", async () => {
  const onNodeInputUpdate = jest.fn();
  const res = await runFromNode(
    "n2",
    {
      variables: [
        { id: "world", name: "world", type: "text", description: "", defaultValue: "test1" },
        { id: "hello", name: "hello", type: "text", description: "", defaultValue: "test2" },
      ],
    },
    simpleChain,
    {
      onNodeInputUpdate,
      onNodeRunError: () => {},
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(2);
  expect(res).toMatchSnapshot();
});
