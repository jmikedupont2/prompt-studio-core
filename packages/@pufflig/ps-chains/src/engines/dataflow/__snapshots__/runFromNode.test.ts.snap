// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`avoid hanging on loops 1`] = `
{
  "n1": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
}
`;

exports[`can handle input from multiple origins 1`] = `
{
  "n1": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
  "n2": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
  "n3": {
    "input": {
      "data1": "Hello {{World}}",
      "data2": "Hello {{World}}",
    },
  },
}
`;

exports[`can resolve variables correctly 1`] = `
{
  "n1": {
    "input": {
      "data": "\${{ps:ref:file:MY_FILE}} {{keep}}",
    },
    "status": "idle",
  },
  "n2": {
    "input": {
      "data": "INSERTED_FILE {{keep}}",
    },
  },
}
`;

exports[`correctly map output from one node onto another 1`] = `
{
  "n1": {
    "input": {
      "data": "test data",
    },
  },
  "n2": {
    "input": {
      "data1": "test data",
      "data2": "",
    },
  },
  "n3": {
    "input": {
      "data": "test data",
    },
  },
}
`;

exports[`ignore autorun=false flag on root node 1`] = `
{
  "n1": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
  "n2": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
}
`;

exports[`ignore inputs from nested unreachable nodes 1`] = `
{
  "n1": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
  "n2": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
  "n4": {
    "input": {
      "data1": "Hello {{World}}",
      "data2": "",
    },
  },
  "n5": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
}
`;

exports[`ignore inputs from unreachable nodes 1`] = `
{
  "n1": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
  "n2": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
  "n3": {
    "input": {
      "data1": "Hello {{World}}",
      "data2": "",
    },
  },
  "n4": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
}
`;

exports[`merge existing state on input 1`] = `
{
  "n1": {
    "input": {
      "data1": "IS_OVERWRITTEN",
      "data2": "KEEP",
    },
    "status": "idle",
  },
}
`;

exports[`nodes with autorun=false down the chain do not update the inputs of their targets 1`] = `
{
  "n1": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
  "n2": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
}
`;

exports[`only run a node once all inputs are resolved 1`] = `
{
  "n1": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
  "n2": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
  "n3": {
    "input": {
      "data1": "Hello {{World}}",
      "data2": "Hello {{World}}",
    },
  },
  "n4": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
}
`;

exports[`run a single node chain 1`] = `
{
  "n1": {
    "input": {
      "data": "Hello World",
    },
    "status": "idle",
  },
}
`;

exports[`running a node updates the input of its children 1`] = `
{
  "n1": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
  "n2": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
}
`;

exports[`should throw error if the node is undefined 1`] = `[Error: Definition for node not_a_node not found]`;

exports[`update array variables 1`] = `
{
  "n2": {
    "input": {
      "data": "",
      "variables": [
        {
          "defaultValue": "test1",
          "description": "",
          "id": "world",
          "name": "world",
          "type": "text",
        },
        {
          "defaultValue": "test2",
          "description": "",
          "id": "hello",
          "name": "hello",
          "type": "text",
        },
      ],
    },
  },
}
`;
